### ISSUE TITLE
Improve dynamic configuration of envoy proxies

### ISSUE THREAD

**Issue by aojea** [kubernetes-sigs/cloud-provider-kind#69]
Existing dynamic configuration implementation is based on Dynamic Configuration from filesystem
https://www.envoyproxy.io/docs/envoy/latest/start/quick-start/configuration-dynamic-filesystem

There is another dynamic configuration option based on a control plane server, but AFAIK this only works with envoys connection to the control plane, something that is not feasible on Macs and Windows as the containers run in VMs https://www.envoyproxy.io/docs/envoy/latest/start/quick-start/configuration-dynamic-control-plane

I'd prefer avoid to build our own images, is because of that we override the initial configuration at the container start using the official envoy image (this is executed as the envoy user)

https://github.com/kubernetes-sigs/cloud-provider-kind/blob/ad8f47301724c7615b5baef24fa7fce82d364f33/pkg/loadbalancer/server.go#L220-L225

but this present problems on container restarts :/ , something we need to deal with because of this bug in envoy https://github.com/kubernetes-sigs/cloud-provider-kind/pull/68

So, basically current implementation copies the new config files to the container and then performs a rename so envoy picks the new config (this is executed as root)

https://github.com/kubernetes-sigs/cloud-provider-kind/blob/ad8f47301724c7615b5baef24fa7fce82d364f33/pkg/loadbalancer/proxy.go#L255-L279

There are several places I can see we can do better:

...

```[tasklist]
### Tasks
- [ ] Reduce the number of container exec calls
- [ ] Improve the resilience of the solution by making the commands idempotent
- [ ] Don't update files if the config didn't change
```

**Comment by BenTheElder** [comment]
> So, basically current implementation copies the new config files to the container and then performs a rename so envoy picks the new config (this is executed as root)

That's a very normal way to do things and a reasonable solution IMHO, this is how kubelet writes configmaps.

I don't think we should depend on a control plane endpoint.

**Comment by BenTheElder** [comment]
> but this present problems on container restarts :/

Such as? We should be able to persist the config across restarts.

**Comment by aojea** [comment]
> Such as? We should be able to persist the config across restarts.

I had to do this chmod https://github.com/kubernetes-sigs/cloud-provider-kind/blob/ad8f47301724c7615b5baef24fa7fce82d364f33/pkg/loadbalancer/proxy.go#L274

**Comment by BenTheElder** [comment]
There are other options, such as exec-ing as the same user as envoy. The chmod doesn't seem like a big problem anyhow though...?

**Comment by aojea** [comment]
> There are other options, such as exec-ing as the same user as envoy. The chmod doesn't seem like a big problem anyhow though...?

yeah, I may be overthinking the problem, mainly because I saw some traces on some error failures about the command executed, but may be a red herring ... I will keep this open until I have time again to investigate more

**Comment by aojea** [comment]
Ok, it seems the commands are idempotent now , a bit ugly but it is ok, and that envoy is smart enough to detect if something changed to trigger an update

> [2024-05-23 23:30:00.833][1][info][upstream] [source/common/upstream/cds_api_helper.cc:32] cds: add 1 cluster(s), remove 0 cluster(s)
> [2024-05-23 23:30:00.834][1][info][upstream] [source/common/upstream/cds_api_helper.cc:71] cds: added/updated 1 cluster(s), skipped 0 unmodified cluster(s)

we can close
